controllers/authpolicy_envoysecuritypolicy_controller.go

```diff
@@ -0,0 +1,190 @@
+		logger.Error(err, "failed to reconcile envoy SecurityPolicy resource")
+		return err
+	}
+
+	return nil
+}
+
+func envoySecurityPolicy(targetNetworkObject client.Object, kuadrantNamespace string) *egv1alpha1.SecurityPolicy {
+	targetNetworkObjectGvk := targetNetworkObject.GetObjectKind().GroupVersionKind()
+	esp := &egv1alpha1.SecurityPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      EnvoySecurityPolicyName(targetNetworkObject.GetName()),
+			Namespace: targetNetworkObject.GetNamespace(),
+			Labels: map[string]string{
+				kuadrant.KuadrantNamespaceAnnotation: kuadrantNamespace,
+			},
+		},
+		Spec: egv1alpha1.SecurityPolicySpec{
+			TargetRef: gatewayapiv1alpha2.PolicyTargetReferenceWithSectionName{
+				PolicyTargetReference: gatewayapiv1alpha2.PolicyTargetReference{
+					Group:     gatewayapiv1.Group(targetNetworkObjectGvk.Group),
+					Kind:      gatewayapiv1.Kind(targetNetworkObjectGvk.Kind),
+					Name:      gatewayapiv1.ObjectName(targetNetworkObject.GetName()),
+					Namespace: ptr.To(gatewayapiv1.Namespace(targetNetworkObject.GetNamespace())),
+				},
+				SectionName: nil,
+			},
+			ExtAuth: &egv1alpha1.ExtAuth{
+				GRPC: &egv1alpha1.GRPCExtAuthService{
+					BackendRef: gatewayapiv1.BackendObjectReference{
+						Name:      kuadrant.AuthorinoServiceName,
+						Kind:      ptr.To[gatewayapiv1.Kind]("Service"),
+						Namespace: ptr.To(gatewayapiv1.Namespace(kuadrantNamespace)),
+						Port:      ptr.To(gatewayapiv1.PortNumber(50051)),
+					},
+				},
+			},
+		},
+	}
+	kuadrant.AnnotateObject(esp, kuadrantNamespace)
+	return esp
+}
+
+func EnvoySecurityPolicyName(targetName string) string {
+	return fmt.Sprintf("on-%s", targetName)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *AuthPolicyEnvoySecurityPolicyReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	ok, err := kuadrantenvoygateway.IsEnvoyGatewaySecurityPolicyInstalled(mgr.GetRESTMapper())
+	if err != nil {
+		return err
+	}
+	if !ok {
+		r.Logger().Info("EnvoyGateway SecurityPolicy controller disabled. EnvoyGateway API was not found")
+		return nil
+	}
+
+	ok, err = kuadrantgatewayapi.IsGatewayAPIInstalled(mgr.GetRESTMapper())
+	if err != nil {
+		return err
+	}
+	if !ok {
+		r.Logger().Info("EnvoyGateway SecurityPolicy controller disabled. GatewayAPI was not found")
+		return nil
+	}
+
+	httpRouteToParentGatewaysEventMapper := mappers.NewHTTPRouteToParentGatewaysEventMapper(
+		mappers.WithLogger(r.Logger().WithName("httpRouteToParentGatewaysEventMapper")),
+	)
+	apToParentGatewaysEventMapper := mappers.NewPolicyToParentGatewaysEventMapper(
+		mappers.WithLogger(r.Logger().WithName("authpolicyToParentGatewaysEventMapper")),
+		mappers.WithClient(r.Client()),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&gatewayapiv1.Gateway{}).
+		Owns(&egv1alpha1.SecurityPolicy{}).
```

Comment:
missing ownerref on the secpolicy?

Reply:
I'm a little unsure who should be the owner of the security policies here, I realise that given the format of this controller block it suggests it should be the gateway. Given that there is one sec policy per auth policy target the owner seems more likely to be the auth policy, or the target of the policy.

If the owner is the httproute/gateway that is targeted this actually solves one of the issues I'm facing regarding deletion (currently the sec policy remains if there's still an auth policy but the target of the policy is deleted). Does this break the model though if I set the owner of sec policies on route nodes to the route node within the controller for gateways?

Reply:
For now I've set the owner as the target gateway/httproute (which has allowed me to enable the deletion integration tests)